core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala

```diff
@@ -46,6 +46,26 @@ class DirectTaskResult[T](var valueBytes: ByteBuffer, var accumUpdates: Map[Long
 
   def this() = this(null.asInstanceOf[ByteBuffer], null, null)
 
+  override def toString: String = valueBytes.toString + " " + accumUpdates + " " + metrics
+
+  override def equals(other: Any): Boolean = other match {
+    case that: DirectTaskResult[_] => {
+      if (!this.valueBytes.equals(that.valueBytes)) return false
+
+      val accumSize = if (accumUpdates != null) accumUpdates.size else 0
+      val thatAccumSize = if (that.accumUpdates != null) that.accumUpdates.size else 0
+      if (accumSize != thatAccumSize) return false
+      if (accumSize > 0) {
+        val b = this.accumUpdates.keys.forall { key =>
+          this.accumUpdates.get(key) == that.accumUpdates.get(key)
+        }
+        if (!b) return false;
+      }
+      this.metrics.equals(that.metrics)
+    }
```

Comment:
similarly, this can be simplified
```
val accumEquals =
  if (accumUpdates != null) {
    accumUpdates.keys.forall { k => accumUpdates.get(k) == that.accumUpdates.get(k) }
  } else {
    that.accumUpdates == null
  }
  valueBytes.equals(that.valueBytes) &&
  metrics.equals(that.metrics) &&
  accumEquals
```